File: sdk-go/builders/builders_test.go
package builders

import (
	"testing"

	"github.com/ai-sdlc-framework/ai-sdlc/sdk-go/core"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestPipelineBuilder(t *testing.T) {
	p, err := NewPipelineBuilder("test-pipeline").
		Namespace("team-a").
		Label("env", "staging").
		AddTrigger("issue.assigned", nil).
		AddProvider("issueTracker", "linear", nil).
		AddStage("plan", "planner", nil).
		AddStage("code", "coder", &StageOptions{
			QualityGates: []string{"code-quality"},
			Timeout:      "PT30M",
		}).
		WithRouting(&core.RoutingConfig{
			ComplexityThresholds: map[string]core.ComplexityThreshold{
				"simple": {Min: 1, Max: 3, Strategy: "fully-autonomous"},
			},
		}).
		Build()

	require.NoError(t, err)
	assert.Equal(t, "test-pipeline", p.Metadata.Name)
	assert.Equal(t, "team-a", p.Metadata.Namespace)
	assert.Equal(t, core.APIVersion, p.APIVersion)
	assert.Equal(t, "Pipeline", p.Kind)
	assert.Len(t, p.Spec.Triggers, 1)
	assert.Len(t, p.Spec.Stages, 2)
	assert.Equal(t, "planner", p.Spec.Stages[0].Agent)
	assert.Equal(t, "staging", p.Metadata.Labels["env"])
	assert.NotNil(t, p.Spec.Routing)
}

func TestPipelineBuilderValidation(t *testing.T) {
	_, err := NewPipelineBuilder("").Build()
	assert.Error(t, err)

	_, err = NewPipelineBuilder("p").Build()
	assert.Error(t, err, "should fail without triggers")

	_, err = NewPipelineBuilder("p").AddTrigger("x", nil).Build()
	assert.Error(t, err, "should fail without stages")
}

func TestAgentRoleBuilder(t *testing.T) {
	a, err := NewAgentRoleBuilder("coder-agent").
		Role("Senior Software Engineer").
		Goal("Write clean code").
		Backstory("Expert in Go and Python").
		AddTool("git").
		AddTool("vscode").
		WithConstraints(&core.Constraints{RequireTests: boolPtr(true)}).
		AddHandoff("reviewer", "code-complete", nil).
		AddSkill("go-development", "Expert Go development", []string{"go", "backend"}).
		Build()

	require.NoError(t, err)
	assert.Equal(t, "coder-agent", a.Metadata.Name)
	assert.Equal(t, "Senior Software Engineer", a.Spec.Role)
	assert.Len(t, a.Spec.Tools, 2)
	assert.Len(t, a.Spec.Handoffs, 1)
	assert.Len(t, a.Spec.Skills, 1)
}

func TestAgentRoleBuilderValidation(t *testing.T) {
	_, err := NewAgentRoleBuilder("a").Goal("g").AddTool("t").Build()
	assert.Error(t, err, "should fail without role")

	_, err = NewAgentRoleBuilder("a").Role("r").AddTool("t").Build()
	assert.Error(t, err, "should fail without goal")

	_, err = NewAgentRoleBuilder("a").Role("r").Goal("g").Build()
	assert.Error(t, err, "should fail without tools")
}

func TestQualityGateBuilder(t *testing.T) {
	threshold := 80.0
	q, err := NewQualityGateBuilder("code-quality").
		WithScope(&core.GateScope{AuthorTypes: []string{"ai-agent"}}).
		AddGate("coverage", core.EnforcementHardMandatory,
			core.GateRule{Metric: "coverage", Operator: ">=", Threshold: &threshold}, nil).
		Build()

	require.NoError(t, err)
	assert.Equal(t, "code-quality", q.Metadata.Name)
	assert.Len(t, q.Spec.Gates, 1)
	assert.NotNil(t, q.Spec.Scope)
}

func TestAutonomyPolicyBuilder(t *testing.T) {
	a, err := NewAutonomyPolicyBuilder("progressive").
		AddLevel(core.AutonomyLevel{
			Level:       0,
			Name:        "Intern",
			Permissions: core.Permissions{Read: []string{"*"}, Write: []string{"draft-pr"}, Execute: []string{"test"}},
			Guardrails:  core.Guardrails{RequireApproval: core.ApprovalAll},
			Monitoring:  core.MonitoringContinuous,
		}).
		AddPromotionCriteria("0-to-1", core.PromotionCriteria{
			MinimumTasks:      10,
			Conditions:        []core.MetricCondition{{Metric: "approval-rate", Operator: ">=", Threshold: 0.9}},
			RequiredApprovals: []string{"tech-lead"},
		}).
		AddDemotionTrigger(core.DemotionTrigger{
			Trigger: "security-violation", Action: core.DemoteToZero, Cooldown: "7d",
		}).
		Build()

	require.NoError(t, err)
	assert.Equal(t, "progressive", a.Metadata.Name)
	assert.Len(t, a.Spec.Levels, 1)
}

func TestAdapterBindingBuilder(t *testing.T) {
	a, err := NewAdapterBindingBuilder("linear-binding").
		Interface(core.InterfaceIssueTracker).
		Type("linear").
		Version("1.0.0").
		Source("registry://adapters/linear").
		Config(map[string]interface{}{"apiUrl": "https://api.linear.app"}).
		WithHealthCheck(&core.HealthCheck{Interval: "30s", Timeout: "5s"}).
		Build()

	require.NoError(t, err)
	assert.Equal(t, "linear-binding", a.Metadata.Name)
	assert.Equal(t, core.InterfaceIssueTracker, a.Spec.Interface)
	assert.NotNil(t, a.Spec.HealthCheck)
}

func TestAdapterBindingBuilderValidation(t *testing.T) {
	_, err := NewAdapterBindingBuilder("a").Type("t").Version("1.0.0").Build()
	assert.Error(t, err, "should fail without interface")
}

func TestDistributionManifest(t *testing.T) {
	m, err := BuildDistribution("my-dist", "1.0.0", []ManifestEntry{
		{Path: "pipeline.yaml", Kind: "Pipeline", Name: "my-pipeline"},
	})
	require.NoError(t, err)
	assert.Equal(t, "my-dist", m.Metadata.Name)
	assert.Len(t, m.Resources, 1)
}

func TestParseManifestYAML(t *testing.T) {
	yaml := `
apiVersion: ai-sdlc.io/v1alpha1
kind: Distribution
metadata:
  name: test-dist
resources:
  - path: pipeline.yaml
    kind: Pipeline
    name: my-pipeline
`
	m, err := ParseBuilderManifest([]byte(yaml))
	require.NoError(t, err)
	assert.Equal(t, "test-dist", m.Metadata.Name)
}

func boolPtr(b bool) *bool { return &b }

File: dogfood/src/orchestrator/builders.test.ts
import { describe, it, expect } from 'vitest';
import {
  buildDogfoodPipeline,
  buildDogfoodAgentRole,
  buildDogfoodQualityGate,
  buildDogfoodAutonomyPolicy,
  buildDogfoodAdapterBinding,
  PipelineBuilder,
  AgentRoleBuilder,
  QualityGateBuilder,
  AutonomyPolicyBuilder,
  AdapterBindingBuilder,
  parseBuilderManifest,
  validateBuilderManifest,
  parsePipelineManifest,
  buildPipelineDistribution,
  API_VERSION,
} from './builders.js';

describe('Resource builders', () => {
  describe('buildDogfoodPipeline()', () => {
    it('creates a valid pipeline resource', () => {
      const pipeline = buildDogfoodPipeline();
      expect(pipeline.kind).toBe('Pipeline');
      expect(pipeline.apiVersion).toBe('ai-sdlc.io/v1alpha1');
      expect(pipeline.metadata.name).toBe('dogfood-pipeline');
      expect(pipeline.spec.stages.length).toBeGreaterThan(0);
      expect(pipeline.spec.triggers.length).toBeGreaterThan(0);
    });

    it('includes routing configuration', () => {
      const pipeline = buildDogfoodPipeline();
      expect(pipeline.spec.routing).toBeDefined();
      expect(pipeline.spec.routing!.complexityThresholds).toBeDefined();
      expect(Object.keys(pipeline.spec.routing!.complexityThresholds!).length).toBe(4);
    });

    it('includes providers', () => {
      const pipeline = buildDogfoodPipeline();
      expect(pipeline.spec.providers).toBeDefined();
      expect(pipeline.spec.providers['issueTracker']).toBeDefined();
      expect(pipeline.spec.providers['sourceControl']).toBeDefined();
    });
  });

  describe('buildDogfoodAgentRole()', () => {
    it('creates a valid agent role resource', () => {
      const role = buildDogfoodAgentRole();
      expect(role.kind).toBe('AgentRole');
      expect(role.metadata.name).toBe('code-agent');
      expect(role.spec.role).toBe('AI Software Engineer');
      expect(role.spec.tools.length).toBeGreaterThan(0);
    });

    it('includes constraints and handoffs', () => {
      const role = buildDogfoodAgentRole();
      expect(role.spec.constraints).toBeDefined();
      expect(role.spec.constraints!.maxFilesPerChange).toBe(15);
      expect(role.spec.handoffs!.length).toBe(1);
    });

    it('includes skills', () => {
      const role = buildDogfoodAgentRole();
      expect(role.spec.skills!.length).toBe(1);
      expect(role.spec.skills![0].id).toBe('typescript-development');
    });
  });

  describe('buildDogfoodQualityGate()', () => {
    it('creates a valid quality gate resource', () => {
      const gate = buildDogfoodQualityGate();
      expect(gate.kind).toBe('QualityGate');
      expect(gate.metadata.name).toBe('issue-quality');
      expect(gate.spec.gates.length).toBe(3);
    });

    it('includes scope configuration', () => {
      const gate = buildDogfoodQualityGate();
      expect(gate.spec.scope).toBeDefined();
      expect(gate.spec.scope!.authorTypes).toContain('ai-agent');
    });
  });

  describe('buildDogfoodAutonomyPolicy()', () => {
    it('creates a valid autonomy policy resource', () => {
      const policy = buildDogfoodAutonomyPolicy();
      expect(policy.kind).toBe('AutonomyPolicy');
      expect(policy.metadata.name).toBe('progressive-trust');
      expect(policy.spec.levels.length).toBe(3);
    });

    it('includes promotion criteria and demotion triggers', () => {
      const policy = buildDogfoodAutonomyPolicy();
      expect(Object.keys(policy.spec.promotionCriteria).length).toBeGreaterThan(0);
      expect(policy.spec.demotionTriggers.length).toBeGreaterThan(0);
    });
  });

  describe('buildDogfoodAdapterBinding()', () => {
    it('creates a valid adapter binding resource', () => {
      const binding = buildDogfoodAdapterBinding();
      expect(binding.kind).toBe('AdapterBinding');
      expect(binding.metadata.name).toBe('github-adapter');
      expect(binding.spec.interface).toBe('IssueTracker');
      expect(binding.spec.type).toBe('github');
    });

    it('includes health check', () => {
      const binding = buildDogfoodAdapterBinding();
      expect(binding.spec.healthCheck).toBeDefined();
    });
  });

  describe('fluent builder APIs', () => {
    it('PipelineBuilder supports method chaining', () => {
      const p = new PipelineBuilder('test')
        .label('env', 'test')
        .annotation('note', 'test')
        .addStage({ name: 's1', agent: 'a1' })
        .build();
      expect(p.metadata.labels!['env']).toBe('test');
    });

    it('AgentRoleBuilder supports method chaining', () => {
      const r = new AgentRoleBuilder('test', 'role', 'goal')
        .addTool('Read')
        .addTool('Write')
        .build();
      expect(r.spec.tools).toEqual(['Read', 'Write']);
    });

    it('QualityGateBuilder supports method chaining', () => {
      const g = new QualityGateBuilder('test')
        .addGate({
          name: 'g1',
          enforcement: 'advisory',
          rule: { metric: 'm', operator: '>=', threshold: 1 },
        })
        .build();
      expect(g.spec.gates).toHaveLength(1);
    });

    it('AutonomyPolicyBuilder supports method chaining', () => {
      const p = new AutonomyPolicyBuilder('test')
        .addLevel({
          level: 0,
          name: 'supervised',
          permissions: { read: ['**'], write: [], execute: [] },
          guardrails: { requireApproval: 'all', maxLinesPerPR: 100 },
          monitoring: 'continuous',
        })
        .build();
      expect(p.spec.levels).toHaveLength(1);
    });

    it('AdapterBindingBuilder supports method chaining', () => {
      const b = new AdapterBindingBuilder('test', 'IssueTracker', 'github', '1.0.0')
        .config({ org: 'test' })
        .build();
      expect(b.spec.config).toEqual({ org: 'test' });
    });
  });

  describe('API_VERSION', () => {
    it('matches the expected version string', () => {
      expect(API_VERSION).toBe('ai-sdlc.io/v1alpha1');
    });

    it('is used in built resources', () => {
      const pipeline = buildDogfoodPipeline();
      expect(pipeline.apiVersion).toBe(API_VERSION);
    });
  });

  describe('distribution builder', () => {
    it('parseBuilderManifest parses YAML', () => {
      const yaml = `
spec_version: "1.0"
adapters: []
output:
  format: yaml
  directory: ./out
`;
      const manifest = parseBuilderManifest(yaml);
      expect(manifest.spec_version).toBe('1.0');
    });

    it('validateBuilderManifest returns validation result', () => {
      const manifest = parseBuilderManifest(`
spec_version: "1.0"
adapters: []
output:
  format: yaml
  directory: ./out
`);
      const result = validateBuilderManifest(manifest);
      expect(typeof result.valid).toBe('boolean');
    });
  });

  describe('parsePipelineManifest()', () => {
    it('parses a valid manifest YAML and returns the manifest object', () => {
      const yaml = `
spec_version: "1.0"
adapters:
  - name: test-adapter
    version: "1.0.0"
    source: "local:./adapters/test"
output:
  name: test-dist
  version: "1.0.0"
  format: yaml
  directory: ./out
`;
      const manifest = parsePipelineManifest(yaml);
      expect(manifest.spec_version).toBe('1.0');
      expect(manifest.adapters).toHaveLength(1);
      expect(manifest.adapters[0].name).toBe('test-adapter');
    });

    it('throws when the manifest is invalid (no adapters)', () => {
      const yaml = `
spec_version: "1.0"
adapters: []
output:
  name: test
  version: "1.0.0"
  format: yaml
  directory: ./out
`;
      expect(() => parsePipelineManifest(yaml)).toThrow('Invalid manifest');
    });

    it('throws when the manifest is missing spec_version', () => {
      const yaml = `
adapters:
  - name: a
    version: "1.0.0"
    source: "local:./a"
output:
  name: test
  version: "1.0.0"
  format: yaml
  directory: ./out
`;
      // parseBuilderManifest throws directly for missing spec_version
      expect(() => parsePipelineManifest(yaml)).toThrow();
    });
  });

  describe('buildPipelineDistribution()', () => {
    it('builds a distribution from a valid manifest', async () => {
      const yaml = `
spec_version: "1.0"
adapters:
  - name: test-adapter
    version: "1.0.0"
    source: "local:./adapters/test"
output:
  name: test-dist
  version: "1.0.0"
  format: yaml
  directory: ./out
`;
      const manifest = parsePipelineManifest(yaml);
      const result = await buildPipelineDistribution(manifest);
      expect(result).toBeDefined();
      expect(typeof result.valid).toBe('boolean');
    });

    it('returns a result object with expected shape', async () => {
      const yaml = `
spec_version: "1.0"
adapters:
  - name: my-adapter
    version: "2.0.0"
    source: "local:./adapters/my"
output:
  name: my-dist
  version: "2.0.0"
  format: yaml
  directory: ./dist
`;
      const manifest = parsePipelineManifest(yaml);
      const result = await buildPipelineDistribution(manifest, {});
      expect(result).toHaveProperty('valid');
      expect(result).toHaveProperty('manifest');
      expect(result).toHaveProperty('resolved');
      expect(result).toHaveProperty('errors');
      expect(result).toHaveProperty('warnings');
    });
  });
});
